### PR TITLE
Enhance: Enable usage of white spaces in project's path

### PR DESCRIPTION
**Pull request by radurentea**
## Description

Enable usage of white spaces in project's path.
https://jira.espressif.com:8443/browse/VSC-808

## Type of change

Please delete options that are not relevant.
- New feature (non-breaking change which adds functionality)
- This change requires a documentation update

## Steps to test this pull request

Provide a list of steps to test changes in this PR and required output
1. Make sure all the above have whitespaces in their paths:

- ESP-IDF folder
- ESP-IDF Tools folder
- Your project
>Note: For ESP-IDF folder and ESP-IDF Tools folder it means you will need to make a new installation for them.
2. Create a project that has a white space in it. Ex: "My Project"
3. Build, Flash, Monitor it.
>Note: Needs to be tested for all variations of: **flashtype method** (dfu, jtag, uart), **Default Terminal in VsCode**:
Windows: Command Prompt and Powershell
Mac: zsh/bash
Linux: bash
 (open command palette in VS Code [CTRL + SHIFT +P] and use the command: `Terminal: Select Default Profile` 
6. There should be no output error regarding the paths. Such as 'invalid path' or 'invalid argument' etc.

- Expected behaviour:
With these changes, user should be able to use the VsCode extension and it's features on projects that have white spaces in their path or in the ESP-IDF and ESP-IDF Tools installation paths

## How has this been tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- Tested building, flashing and monitoring on all variations of Terminals (Windows: Powershell, Mac: zsh/bash, Linux: bash) and flashing methods (DFU, JTAG, UART).

**Test Configuration**:
* ESP-IDF Version: 5.1
* OS (Windows,Linux and macOS): Windows, Linux, macOS

## Checklist
- [x] PR Self Reviewed
- [x] Applied Code formatting
- [x] Added Documentation
- [ ] Added Unit Test
- [x] Verified on all platforms - Windows,Linux and macOS

### CODE REVIEW

**Review by brianignacio5 — CHANGES_REQUESTED** [review]
Left a few changes about code. Will test on Windows soon.

File: src/espIdf/monitor/index.ts [hunk 1]
```diff
@@ -16,7 +16,8 @@
  * limitations under the License.
  */
 
-import { appendIdfAndToolsToPath } from "../../utils";
+import { get } from "http";
```
[the remark this answers is not in the document]
Comment (radurentea) [reply src/espIdf/monitor/index.ts:19]: Done, thank you!

File: src/espIdf/size/idfSizeTask.ts [hunk 2]
```diff
@@ -30,8 +30,9 @@ import {
 } from "vscode";
 import { readParameter } from "../../idfConfiguration";
 import { TaskManager } from "../../taskManager";
-import { appendIdfAndToolsToPath } from "../../utils";
+import { appendIdfAndToolsToPath, getUserShell } from "../../utils";
 import { getProjectName } from "../../workspaceConfig";
+import { get } from "http";
```
[the remark this answers is not in the document]
Comment (radurentea) [reply src/espIdf/size/idfSizeTask.ts:35]: Done thank you!

**Comment by brianignacio5** [comment]
I did a series of tests in Windows 10.

While the current PR does indeed solve the issue of spaces in the current project path for build, flash monitor commands; it doesn't take into account the support of spaces in ESP-IDF or ESP-IDF tools which could also happen. The validation in the setup wizard was removed to validate spaces, so I think we should also support spaces in ESP-IDF, ESP-IDF tools and PATH. 

I think we should have some ESP-IDF version validation too, because spaces in ESP-IDF started to be supported from v5.0 if I'm not mistaken.

**Comment by radurentea** [comment]
> While the current PR does indeed solve the issue of spaces in the current project path for build, flash monitor commands; it doesn't take into account the support of spaces in ESP-IDF or ESP-IDF tools which could also happen. The validation in the setup wizard was removed to validate spaces, so I think we should also support spaces in ESP-IDF, ESP-IDF tools and PATH.

But it does offer support for spaces in esp-idf and esp-idf tools. If esp-idf and esp-idf tools paths contain spaces, they should still work, that's how I've tested the changes for all OS.

> 
> I think we should have some ESP-IDF version validation too, because spaces in ESP-IDF started to be supported from v5.0 if I'm not mistaken.

Where should I add this validation? I was thinking to maybe add like a warning inside the UI of ESP-IDF SETUP web view for express and advanced (if user selects an esp-idf version that doesn't support white spaces) and for the Use Existing Setup (should have like a warning inside each option again that uses an esp-idf version that doesn't support white spaces)
What do you think?

**Comment by brianignacio5** [comment]
> > While the current PR does indeed solve the issue of spaces in the current project path for build, flash monitor commands; it doesn't take into account the support of spaces in ESP-IDF or ESP-IDF tools which could also happen. The validation in the setup wizard was removed to validate spaces, so I think we should also support spaces in ESP-IDF, ESP-IDF tools and PATH.
> 
> But it does offer support for spaces in esp-idf and esp-idf tools. If esp-idf and esp-idf tools paths contain spaces, they should still work, that's how I've tested the changes for all OS.
> 
> > I think we should have some ESP-IDF version validation too, because spaces in ESP-IDF started to be supported from v5.0 if I'm not mistaken.
> 
> Where should I add this validation? I was thinking to maybe add like a warning inside the UI of ESP-IDF SETUP web view for express and advanced (if user selects an esp-idf version that doesn't support white spaces) and for the Use Existing Setup (should have like a warning inside each option again that uses an esp-idf version that doesn't support white spaces) What do you think?

Yes having the warning in the Setup Wizard and adding in the readme and install tutorial a warning for all esp-idf < 5.0 should be enough. I just mentioned the version because we still have to support v4.3 and v4.4 until end of next year

**Comment by birdistheword96** [comment]
Following this intently as its a really important fix. I can't change my user-name (which has a space in) at work so I currently cannot use VS-Code extension, I have to use the power-shell and call the commands separately. Cannot wait for this fix to be pushed.

File: src/views/setup/components/selectEspIdf.vue [hunk 3]
```diff
@@ -137,6 +152,28 @@ export default class SelectEspIdf extends Vue {
   set showGithubTags(showTags: boolean) {
     this.setShowIdfTagList(showTags);
   }
+  get isVersionLowerThan5() {
+    if (this.selectedIdfVersion && this.selectedIdfVersion.name) {
+      // Regular expression to match the version number in the format vX.X.X or release/vX.X
+      const match = this.selectedIdfVersion.name.match(/v(\d+(\.\d+)?(\.\d+)?)/);
```
Comment (brianignacio5) [inline src/views/setup/components/selectEspIdf.vue:158]: Any reason why not using the version validation in the helper function in `.../store/index.ts` ?

**Review by brianignacio5 — APPROVED** [review]
LGTM with minor comment

**Review by brianignacio5 — APPROVED** [review]
LGTM

**Comment by AndriiFilippov** [comment]
@radurentea hi! 

Tested under:
OS - Windows 10 
ESP-IDF: v5.1 

Is there any space set restrictions ? Because when I add more spaces - I can't build project via VS Code, but able to build via terminal

SV Code LOGS: 

```
*  Executing task: cmake -G Ninja -DPYTHON_DEPS_CHECKED=1 -DESP_PLATFORM=1 -B 'c:\Users\AndriiFilippov\project                  name\build' -S 'c:\Users\AndriiFilippov\project                  name' 

CMake Error: The source directory "C:/Users/AndriiFilippov/project name" does not exist.
Specify --help for usage, or press the help button on the CMake GUI.

 *  The terminal process "C:\WINDOWS\System32\WindowsPowerShell\v1.0\powershell.exe -Command cmake -G Ninja -DPYTHON_DEPS_CHECKED=1 -DESP_PLATFORM=1 -B 'c:\Users\AndriiFilippov\project                  name\build' -S 'c:\Users\AndriiFilippov\project                  name'" terminated with exit code: 1.
```


Terminal LOGS:

```
ifi_provisioning.dir/proto-c/wifi_constants.pb-c.c.obj←[883/884] Generating binary image from built executable←[Kesptool.py v4.7.dev1.obj←[K.obj←[K/src/scheme_softap.c.obj←[K
Creating esp32 image...
Merged 2 ELF sections
Successfully created esp32 image.
Generated C:/Users/AndriiFilippov/e/TEST SPACES/esp-idf-v5.1/examples/project                        name/build/project.bin
[884/884] cmd.exe /C "cd /D "C:\Users\AndriiFilippov\e\TES...es/project                        name/build/project.bin""←[Kproject.bin binary size 0x2aa60 bytes. Smallest app partition is 0x100000 bytes. 0xd55a0 bytes (83%) free.

Project build complete. To flash, run this command:
C:\Users\AndriiFilippov\.espressif\python_env\idf5.1_py3.11_env\Scripts\python.exe ..\..\components\esptool_py\esptool\esptool.py -p (PORT) -b 460800 --before default_reset --after hard_reset --chip esp32  write_flash --flash_mode dio --flash_size 2MB --flash_freq 40m 0x1000 build\bootloader\bootloader.bin 0x8000 build\partition_table\partition-table.bin 0x10000 build\project.bin
or run 'idf.py -p (PORT) flash'
```

**Comment by AndriiFilippov** [comment]
@radurentea 

Also whenever I create project with spaces I do have this messages in console: 

```
[proc] Executing command: cmake --version
[proc] The command: cmake --version failed with error: Error: spawn cmake ENOENT
[proc] Executing command: cmake -E capabilities
[proc] The command: cmake -E capabilities failed with error: Error: spawn cmake ENOENT
[variant] Loaded new set of variants
[kit] Successfully loaded 1 kits from C:\Users\AndriiFilippov\AppData\Local\CMakeTools\cmake-tools-kits.json
[main] Configuring project: project is name 
[main] Configuring project: project is name
```

**Comment by radurentea** [comment]
Hi @AndriiFilippov ,

Can you re-test again for both of the issues you've found previously?

**Comment by AndriiFilippov** [comment]
> @radurentea
> 
> Also whenever I create project with spaces I do have this messages in console:
> 
> ```
> [proc] Executing command: cmake --version
> [proc] The command: cmake --version failed with error: Error: spawn cmake ENOENT
> [proc] Executing command: cmake -E capabilities
> [proc] The command: cmake -E capabilities failed with error: Error: spawn cmake ENOENT
> [variant] Loaded new set of variants
> [kit] Successfully loaded 1 kits from C:\Users\AndriiFilippov\AppData\Local\CMakeTools\cmake-tools-kits.json
> [main] Configuring project: project is name 
> [main] Configuring project: project is name
> ```
> 
> Do you know what is it ?

@radurentea , This one still persist, but I think we can create separate ticket for this one ? what do you think ? 

Tested on Linux as well, everything works fine, but on the Extension Configuration finish page I do see this info message: 
image


But, this command does not work with spaces in the PATH : 
image


I think we should do something with this one.

**Comment by radurentea** [comment]
@AndriiFilippov 
Made the modifications for the linux command. You can re-test it.

Regarding this one:
```
[proc] Executing command: cmake --version
[proc] The command: cmake --version failed with error: Error: spawn cmake ENOENT
[proc] Executing command: cmake -E capabilities
[proc] The command: cmake -E capabilities failed with error: Error: spawn cmake ENOENT
[variant] Loaded new set of variants
[kit] Successfully loaded 1 kits from C:\Users\AndriiFilippov\AppData\Local\CMakeTools\cmake-tools-kits.json
[main] Configuring project: project is name 
[main] Configuring project: project is name
```
it doesn't seem to be an issue with current PR, I was not able to reproduce it.

**Comment by AndriiFilippov** [comment]
@radurentea 
Screenshot 2023-09-01 at 10 09 09

Also can't build on MacOS. 
Please, check


``` *  Executing task: cmake -G Ninja -DPYTHON_DEPS_CHECKED=1 -DESP_PLATFORM=1 -B="/Users/andrii/project-name            s/build" -S="/Users/andrii/project-name            s" 

zsh:1: command not found: cmake

 *  The terminal process "/bin/zsh '-l', '-c', 'cmake -G Ninja -DPYTHON_DEPS_CHECKED=1 -DESP_PLATFORM=1 -B="/Users/andrii/project-name            s/build" -S="/Users/andrii/project-name            s"'" failed to launch (exit code: 127). 
```

**Comment by nrgbistro** [comment]
Unfortunately I don't have the time or skill to contribute but I would love to see this merged in the next release!

For context, I put both my `IDF_PATH` and `IDF_TOOLS_PATH` in locations that contain no spaces. However, I still cannot build my project that is stored in the following path: `C:\Users\username\OneDrive - xxxx University\Classes\2024 Spring\IOT\Labs\Lab 1`

Here's the build error:

```
Executing task: cmake -G Ninja -DPYTHON_DEPS_CHECKED=1 -DESP_PLATFORM=1 -B c:\Users\username\OneDrive - xxxx University\Classes\2024 Spring\IOT\Labs\Lab 1\build -S c:\Users\username\OneDrive - xxxx University\Classes\2024 Spring\IOT\Labs\Lab 1 

CMake Warning:
  Ignoring extra path from command line:

   "1"


CMake Error: Unknown argument -
CMake Error: Run 'cmake --help' for all supported options.
```

I've tried a few combinations of paths and can't seem to get consistent results unless there are no spaces present in the path, which makes using OneDrive very difficult. I'm using v1.7.0 of this extension.

**Comment by radurentea** [comment]
New PR for adding support for white spaces can be found here:
https://github.com/espressif/vscode-esp-idf-extension/pull/1159

**Pull request closed without merging**